File: src/main/com/natpryce/snodge/mutagens/ReplaceJsonElement.java
package com.natpryce.snodge.mutagens;

import com.google.gson.JsonElement;
import com.natpryce.snodge.DocumentMutation;
import com.natpryce.snodge.JsonPath;
import com.natpryce.snodge.Mutagen;
import kotlin.sequences.Sequence;

import static kotlin.sequences.SequencesKt.sequenceOf;

public class ReplaceJsonElement implements Mutagen {
    private final JsonElement replacement;

    public ReplaceJsonElement(JsonElement replacement) {
        this.replacement = replacement;
    }

    @Override
    public Sequence<DocumentMutation> potentialMutations(JsonElement document, JsonPath pathToElement, JsonElement elementToMutate) {
        return sequenceOf(pathToElement.map(e -> replacement));
    }
}

File: src/main/com/natpryce/snodge/mutagens/ReplaceJsonElement.kt
package com.natpryce.snodge.mutagens

import com.google.gson.JsonElement
import com.natpryce.snodge.JsonNodeMutagen
import com.natpryce.snodge.JsonPath

class ReplaceJsonElement(private val replacement: JsonElement) : JsonNodeMutagen {
    override fun invoke(document: JsonElement, pathToElement: JsonPath, elementToMutate: JsonElement): Sequence<Lazy<JsonElement>> {
        return sequenceOf(lazy { pathToElement.replace(document, replacement) })
    }
}
